### PR TITLE
Added Support to Audit  OIDC Provider/Providers

### PR DESCRIPTION
**Pull request by Nirbhay1997**
### Description

Added Support to Audit OIDC Provider/Providers

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [x] New functionality includes integration tests/controls
- [x] New Terraform resources
- [x] Documentation provided or updated for resources 
- [x] All Integration Tests pass
- [x] All Unit Tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: docs/resources/aws_iam_oidc_provider.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+Use the `aws_iam_oidc_provider` InSpec audit resource to test properties of a single IAM OIDC Provider.
+
+This resource retrieves information about the specified OIDC Provider.
+
+## Syntax
+
+Ensure that a OIDC provider exists.
+
+    describe aws_iam_oidc_provider(open_id_connect_provider_arn: 'open_id_connect_provider_arn') do
```
Comment (soumyo13) [inline docs/resources/aws_iam_oidc_provider.md:16]: Change the value to a meaningful arn.

File: docs/resources/aws_iam_oidc_provider.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      it { should exist }
+    end
+
+## Parameters
+
+`open_id_connect_provider_arn` _(required)_
+
```
Comment (soumyo13) [inline docs/resources/aws_iam_oidc_provider.md:23]: Add the description of open_id_connect_provider_arn

File: docs/resources/aws_iam_oidc_provider.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+| Property | Description|
+| --- | --- |
+| url | The URL that the IAM OIDC provider resource object is associated with. |
+| create_date | The date and time when the IAM OIDC provider resource object was created in the account. |
+| tags | A list of tags that are attached to the specified IAM OIDC provider. |
+
+## Examples
+
+### Ensure a  url is available.
```
Comment (soumyo13) [inline docs/resources/aws_iam_oidc_provider.md:36]:
```suggestion
### Ensure an url is available.
```

File: docs/resources/aws_iam_oidc_providers.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    describe aws_iam_oidc_providers do
+      its('arns') { should include 'open_id_connect_provider_arn' }
+    end
+
+## Matchers
+
+This InSpec audit resource has the following special matchers. For a full list of available matchers, please visit our [Universal Matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+The controls will pass if the `get` method returns at least one result.
```
Comment (soumyo13) [inline docs/resources/aws_iam_oidc_providers.md:41]:
```suggestion
The controls will pass if the `list` method returns at least one result.
```

File: libraries/aws_iam_oidc_provider.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    return nil unless exists?
+    @res[:open_id_connect_provider_arn]
+  end
+
+  def exists?
+    !@res.nil? && !@res.empty?
+  end
+
+  def to_s
+    "open_id_connect_provider_arn: #{@display_name}"
```
Comment (soumyo13) [inline libraries/aws_iam_oidc_provider.rb:38]:
```suggestion
    "Open ID Connect Provider Arn: #{@display_name}"
```

File: libraries/aws_iam_oidc_providers.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    catch_aws_errors do
+      @response = @aws.iam_client.list_open_id_connect_providers
+    end
+    return aws_iam_oidc_providers_rows if !@response || @response.empty?
+    @response.open_id_connect_provider_list.each do |p|
+      aws_iam_oidc_providers_rows += [{
+        arn:     p.arn,
+      }]
+    end
+
```
Comment (soumyo13) [inline libraries/aws_iam_oidc_providers.rb:39]: Remove the empty line

File: libraries/aws_iam_oidc_providers.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    aws_iam_oidc_providers_rows = []
+
+    catch_aws_errors do
+      @response = @aws.iam_client.list_open_id_connect_providers
+    end
+    return aws_iam_oidc_providers_rows if !@response || @response.empty?
+    @response.open_id_connect_provider_list.each do |p|
+      aws_iam_oidc_providers_rows += [{
+        arn:     p.arn,
+      }]
```
Comment (soumyo13) [inline libraries/aws_iam_oidc_providers.rb:37]:
```suggestion
    @response.open_id_connect_provider_list.each do |provider|
      aws_iam_oidc_providers_rows += [{
        arn:     provider.arn,
      }]
```

File: test/integration/build/aws.tf [hunk 8]
```diff
@@ -3883,4 +3883,13 @@ resource "aws_lambda_permission" "allow_cloudwatch" {
...
+}
+resource "aws_iam_openid_connect_provider" "for_oidc" {
+  url = "https://accounts.google.com"
+
+  client_id_list = [
+    "266362248691-342342xasdasdasda-apps.googleusercontent.com",
+  ]
+
+  thumbprint_list = []
 }
```
Comment (soumyo13) [inline test/integration/build/aws.tf:3895]:
```suggestion
}

resource "aws_iam_openid_connect_provider" "for_oidc" {
  url = "https://accounts.google.com"

  client_id_list = [
    "266362248691-342342xasdasdasda-apps.googleusercontent.com",
  ]
  thumbprint_list = []
}
```

File: test/integration/build/outputs.tf [hunk 9]
```diff
@@ -712,3 +712,6 @@ output "aws_secretsmanager_secret_sm_id" {
 output "aws_secretsmanager_secret_sm_arn" {
   value = aws_secretsmanager_secret.aws_secretsmanager_secret_sm_test.arn
 }
+output "oidc_arn" {
+  value = aws_iam_openid_connect_provider.for_oidc.arn
+}
```
Comment (soumyo13) [inline test/integration/build/outputs.tf:717]:
```suggestion
}

output "oidc_arn" {
  value = aws_iam_openid_connect_provider.for_oidc.arn
}
```

File: test/integration/verify/controls/aws_iam_oidc_provider.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+open_id_connect_provider_arn = value("oidc_arn", default: "", description: "")
+
+describe aws_iam_oidc_provider(open_id_connect_provider_arn: 'arn:aws:iam::112758395563:oidc-provider/accounts.google.com') do
```
Comment (soumyo13) [inline test/integration/verify/controls/aws_iam_oidc_provider.rb:4]: Add the control block.

File: test/integration/verify/controls/aws_iam_oidc_providers.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+describe aws_iam_oidc_providers do
```
Comment (soumyo13) [inline test/integration/verify/controls/aws_iam_oidc_providers.rb:3]: Add the control block.

File: test/unit/resources/aws_iam_oidc_provider_test.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+  def test_oidc_provider_exists
+    assert @oidc.exists?
+  end
+
+  def test_open_id_connect_provider_arn
+    assert_equal(@oidc.url, 'test1.com')
+  end
+
+  def test_instance_profile_id
```
Comment (soumyo13) [inline test/unit/resources/aws_iam_oidc_provider_test.rb:41]:
```suggestion
  def test_created_date
```

File: test/unit/resources/aws_iam_oidc_provider_test.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    data[:data] =  mock_parameter
+    data[:client] = Aws::IAM::Client
+    @oidc = AWSIAMOIDCProvider.new(open_id_connect_provider_arn: 'test1', client_args: { stub_responses: true }, stub_data: [data])
+  end
+
+  def test_oidc_provider_exists
+    assert @oidc.exists?
+  end
+
+  def test_open_id_connect_provider_arn
```
Comment (soumyo13) [inline test/unit/resources/aws_iam_oidc_provider_test.rb:37]:
```suggestion
  def test_url
```

File: test/unit/resources/aws_iam_oidc_providers_test.rb [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+  def test_empty_params_ok
+    AWSIAMOIDCProviders.new(client_args: { stub_responses: true })
+  end
+
+  def test_rejects_other_args
+    assert_raises(ArgumentError) { AWSIAMOIDCProviders.new('rubbish') }
+  end
+
+  def test_iam_client_non_existing_for_empty_response
```
Comment (soumyo13) [inline test/unit/resources/aws_iam_oidc_providers_test.rb:15]:
```suggestion
  def test_iam_client_oidc_providers_non_existing_for_empty_response
```

**Review by soumyo13 — APPROVED** [review]
LGTM. Please make the changes.

File: docs/resources/aws_iam_oidc_provider.md (hunk 3 above)
Comment (soumyo13) [reply docs/resources/aws_iam_oidc_provider.md:36]: > 

Done

**Review by IanMadd — APPROVED** [review]

**Pull request merged**